### PR TITLE
feat: scVI-X release

### PR DESCRIPTION
**Pull request by canergen**
Do not merge or review yet.

### CODE REVIEW

**Comment by canergen** [comment]
@ori-kron-wis code can be checked.

File: tests/external/scvix/test_scvix.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    assert model.adata.var.equals(adata_orig.var)
+
+    scvi.settings.seed = 1
+    keys = ["mean", "dispersions", "dropout"]
+    params_latent = model.get_likelihood_parameters(n_samples=200, give_mean=True)
+    for k in keys:
+        assert params_latent[k].shape == params_orig[k].shape
+
+    for k in keys:
+        assert_approx_equal(params_latent[k], params_orig[k])
```
Comment (ori-kron-wis) [inline tests/external/scvix/test_scvix.py:221]: Results are not equal (only tests failure now), I guess, because of resampling of the minified model, you will need to tweak rtol and atol (increase - but how much?)

File: src/scvi/external/scvix/_model.py [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+from ._module import VAEX
+
+if TYPE_CHECKING:
+    from typing import Literal
+
+    import pandas as pd
+    from anndata import AnnData
+
+logger = logging.getLogger(__name__)
+print(2)
```
Comment (ori-kron-wis) [inline src/scvi/external/scvix/_model.py]: typo?

File: src/scvi/external/scvix/_module.py [hunk 3]
```diff
@@ -0,0 +1,718 @@
...
+        unique = torch.unique(assay)
+        pair_penalty = torch.tensor(0.0, device=assay.device)
+        if len(unique) > 1:
+            for i in unique:
+                pp = self.mmd(params, mask=(assay == i))
+                pair_penalty += pp
+
+        return pair_penalty
+
+    def mmd(self, params, mask=None):
```
Comment (ori-kron-wis) [inline src/scvi/external/scvix/_module.py:666]: if mask is None (default), what are mod_1 and mod_2? it will break.
also seems that this is the case cause line 661 invokes mmd with mask==None
its probably not covered in tests

File: src/scvi/external/scvix/_model.py [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+
+
+class SCVIX(
+    EmbeddingMixin,
+    RNASeqMixin,
+    VAEMixin,
+    ArchesMixin,
+    BaseMinifiedModeModelClass,
+):
+    """single-cell Variational Inference :cite:p:`Lopez18`.
```
Comment (ori-kron-wis) [inline src/scvi/external/scvix/_model.py:50]: I guess you will want to put you pre-print title here

File: src/scvi/external/scvix/_model.py [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+        n_cats_per_cov = (
+            self.adata_manager.get_state_registry(REGISTRY_KEYS.CAT_COVS_KEY).n_cats_per_key
+            if REGISTRY_KEYS.CAT_COVS_KEY in self.adata_manager.data_registry
+            else None
+        )
+        self.module = self._module_cls(
+            n_input=self.summary_stats.n_vars,
+            n_batch=self.summary_stats.n_batch,
+            n_assay=self.summary_stats.n_assay,
+            n_adversarial_group=self.summary_stats.get("n_adversarial_group", 1),
```
Comment (ori-kron-wis) [inline src/scvi/external/scvix/_model.py:195]: Shouldn't it be 0 by default to match _trainingplans.py?

File: src/scvi/train/_trainingplans.py [hunk 6]
```diff
@@ -708,11 +744,19 @@ def training_step(self, batch, batch_idx):
...
-            self.manual_backward(loss)
-            opt2.step()
+            loss = 0.0
+            for i in range(self.adversarial_steps):
+                qz = inference_outputs["qz"]
+                z = qz.sample()
+                loss_ = kappa * self.loss_adversarial_classifier(
+                    z, adversarial_group, batch_tensor, True
+                )
+                if i > 1 and (loss - loss_) / loss < 1e-3:
```
Comment (ori-kron-wis) [inline src/scvi/train/_trainingplans.py:754]: if loss -> 0 , this expression goes to inf , was that the intention?

File: tests/external/scvix/test_scvix.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    adata2.obs["batch"] = adata2.obs.batch.cat.rename_categories(["batch_2", "batch_3"])
+    SCVIX.prepare_query_anndata(adata2, dir_path)
+    SCVIX_query = SCVIX.load_query_data(adata2, dir_path)
+    SCVIX_query.train(1, train_size=0.5, plan_kwargs={"weight_decay": 0.0})
+
+    adata3 = SCVIX.prepare_query_anndata(adata2, dir_path, inplace=False)
+    SCVIX_query2 = SCVIX.load_query_data(adata3, dir_path)
+    SCVIX_query2.train(1, train_size=0.5, plan_kwargs={"weight_decay": 0.0})
+
+    # adata4 has more genes and missing 10 genes from adata1
```
Comment (ori-kron-wis) [inline tests/external/scvix/test_scvix.py:164]: should you try to do something with adata4?

File: docs/user_guide/models/scvix.md [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+- **Assay-aware decoding**: Learned gene-level assay biases in the output
+  layer capture technology-specific transcript-detection differences and
+  provide interpretable, reproducible signatures across datasets.
+- **Variational batch embeddings**: Optionally replaces one-hot batch
+  encodings with low-dimensional variational embeddings, reducing parameter
+  count and improving identifiability through a KL regularisation term.
+
+```{topic} Tutorials:
+
+-   {doc}`/tutorials/notebooks/scrna/scvix_tutorial`
```
Comment (ori-kron-wis) [inline docs/user_guide/models/scvix.md:30]: need to add

File: docs/user_guide/models/scvix.md
Comment (ori-kron-wis) [inline docs/user_guide/models/scvix.md:1]: Doc file exists but is unreachable via navigation.
Need to add a link to docs/user_guide/index.md

File: src/scvi/external/scvix/_model.py [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+        :meth:`~scvi.model.base.RNASeqMixin.get_normalized_expression`.
+        """
+        if transform_assay is not None:
+            adata_ = self._validate_anndata(adata)
+            assay_mappings = (
+                self.get_anndata_manager(adata_, required=True)
+                .get_state_registry(REGISTRY_KEYS.ASSAY_KEY)
+                .categorical_mapping
+            )
+            # Hack for transform_assay to be passed in get_normalized_expression.
```
Comment (ori-kron-wis) [inline src/scvi/external/scvix/_model.py:388]: not sure what it is.... double verify (or have a test...)

File: src/scvi/external/scvix/_module.py [hunk 10]
```diff
@@ -0,0 +1,718 @@
...
+            cat_list = n_cats_per_cov_
+        else:
+            cat_list = [n_batch] + n_cats_per_cov_
+        self.encode_assay = encode_assay
+        conditional_category = 0
+        if encode_assay:
+            encode_assay_list = [n_assay]
+        else:
+            encode_assay_list = [0]
+        if not encode_covariates:
```
Comment (ori-kron-wis) [inline src/scvi/external/scvix/_module.py:215]: if encode_covaraites if False, what is self.batch_representation_encoder? might not be utilized later on, but still need a value for this model registry (later save/load etc)

File: src/scvi/utils/_docstrings.py [hunk 11]
```diff
@@ -124,6 +124,12 @@
     integer categories and saved to `adata.obs['_scvi_batch']`. If `None`, assigns the same batch
     to all the data."""
 
+param_assay_key = """\
```
Comment (ori-kron-wis) [inline src/scvi/utils/_docstrings.py:132]: not used...

File: src/scvi/external/scvix/_model.py [hunk 12]
```diff
@@ -0,0 +1,479 @@
...
+            f"gene_likelihood: {gene_likelihood}, "
+            f"batch_representation: {batch_representation}."
+        )
+
+        if prior == "vamp":
+            if pseudoinputs_data_indices is None:
+                pseudoinputs_data_indices = np.random.randint(
+                    0, self.summary_stats.n_cells, n_prior_components
+                )
+            assert pseudoinputs_data_indices.shape[0] == n_prior_components
```
Comment (ori-kron-wis) [inline src/scvi/external/scvix/_model.py:171]: do we need to check assert in the runtime code? It will not tell anything.
 I suggest otherwise to check if condition, and plot an Error message

File: src/scvi/external/scvix/_module.py [hunk 13]
```diff
@@ -0,0 +1,718 @@
...
+                n_latent,
+                n_labels=self.n_labels,
+                use_batch_norm=False,
+                use_layer_norm=True,
+                **cls_parameters,
+            )
+        if prior == "gaussian":
+            self.prior = GaussianPrior()
+        elif prior == "vamp":
+            assert pseudoinput_data is not None, (
```
Comment (ori-kron-wis) [inline src/scvi/external/scvix/_module.py:283]: again, better to add a runtime error/warning message

File: src/scvi/nn/_base_components.py [hunk 14]
```diff
@@ -14,6 +14,44 @@ def _identity(x):
     return x
 
 
+class ConditionalBatchNorm2d(nn.Module):
+    def __init__(self, num_features, num_classes, momentum, eps):
+        super().__init__()
+        self.num_features = num_features
+        self.bn = nn.BatchNorm1d(self.num_features, momentum=momentum, eps=eps, affine=False)
```
Comment (ori-kron-wis) [inline src/scvi/nn/_base_components.py:21]: function name misleading as its says 2D, but runs 1D?

File: src/scvi/train/_trainingplans.py [hunk 15]
```diff
@@ -761,9 +805,7 @@ def configure_optimizers(self):
 
         if self.adversarial_classifier is not False:
```
Comment (ori-kron-wis) [inline src/scvi/train/_trainingplans.py:806]: did we validate TotalVI functionality? multivi? There are parameter changes here (perhaps we can encode for scviX but return to previous values otherwise)

File: src/scvi/train/_trainingplans.py [hunk 16]
```diff
@@ -626,43 +632,68 @@ def __init__(
...
                     stacklevel=settings.warnings_stacklevel,
                 )
                 self.adversarial_classifier = False
             else:
-                self.n_output_classifier = self.module.n_batch
                 self.adversarial_classifier = Classifier(
-                    n_input=self.module.n_latent,
-                    n_hidden=32,
+                    n_input=self.module.n_latent + getattr(self.module, "n_adversarial_group", 0),
+                    n_hidden=128,
```
Comment (ori-kron-wis) [inline src/scvi/train/_trainingplans.py:655]: another parameter change, how di dit affect other adversarial models?

File: src/scvi/external/sysvi/_module.py [hunk 17]
```diff
@@ -100,21 +100,21 @@ def __init__(
 
         self.n_batch = n_batch
         n_cat_list = [n_batch]
-        n_cont = n_continuous_cov
+        n_continuous = n_continuous_cov
```
Comment (ori-kron-wis) [inline src/scvi/external/sysvi/_module.py:103]: not in favor of changing api of other modules.

File: src/scvi/external/sysvi/_base_components.py [hunk 18]
```diff
@@ -59,7 +59,7 @@ def __init__(
         n_input: int,
         n_output: int,
         n_cat_list: list[int],
-        n_cont: int,
+        n_continuous: int,
```
Comment (ori-kron-wis) [inline src/scvi/external/sysvi/_base_components.py:62]: not in favor of changing api of other modules.

File: src/scvi/nn/_base_components.py
Comment (ori-kron-wis) [inline src/scvi/nn/_base_components.py:71]: this is still here, but you changed the API (why?) - my suggestion is to revert to n_cont and update scvi_x and not vice versa

File: src/scvi/nn/_base_components.py [hunk 19]
```diff
@@ -148,20 +205,26 @@ def forward(self, x: torch.Tensor, *cat_list: int, cont: torch.Tensor | None = N
             tensor of values with shape ``(n_in,)``
         cat_list
             list of category membership(s) for this sample
-        cont
-            tensor of continuous covariates with shape ``(n_cont,)``
+        cont_input
```
Comment (ori-kron-wis) [inline src/scvi/nn/_base_components.py:217]: again changing api here that affects other models - not in favor - better to change scvi-x internaly

**Review by ori-kron-wis — COMMENTED** [review]
I think the api changes in core function (trainingplans and base_components) are not really needed and we can avoid most of it by changing it internally in scvix module, thus, without affect other models.

For the changes in adversarial_classifier - it affect other models for sure (totalvi, gimvi) - are we certain they are ok?

File: src/scvi/external/gimvi/_task.py
Comment (ori-kron-wis) [inline src/scvi/external/gimvi/_task.py:1]: a change made to cope with api changes. need to verify

File: src/scvi/external/sysvi/_base_components.py
Comment (ori-kron-wis) [inline src/scvi/external/sysvi/_base_components.py:1]: a change made to cope with api changes. but in this case I think we can revert and change scviX api

File: src/scvi/module/_multivae.py
Comment (ori-kron-wis) [inline src/scvi/module/_multivae.py:1]: a change made to cope with api changes. need to verify

File: src/scvi/external/scvix/_model.py (hunk 9 above)
Comment (canergen) [reply src/scvi/external/scvix/_model.py:388]: Instead of only transform_batch for generation, we want to be able to transform_assay (so conditional generation for another assay)
